implementations/rust/ockam/ockam/src/pipe2/mod.rs

```diff
@@ -0,0 +1,159 @@
+#![allow(unused)] // FIXME
+// If `pipe` is so great, why is there no `pipe2`?
+//! Pipe2 composition system
+//!
+//! Pipe2 offers the ability to compose pipe workers with different
+//! behaviours.  These behaviours are implemented using the
+//! [`SystemHandler`](crate::SystemHandler) abstraction.
+//!
+//! This module is a replacement for [`pipe`](crate::pipe) and should
+//! replace it at some point in the future.
+
+mod hooks;
+mod receiver;
+mod sender;
+
+use crate::{Context, OckamMessage, SystemBuilder, SystemHandler};
+use ockam_core::{
+    compat::{boxed::Box, string::String, vec::Vec},
+    Address, Result, Route,
+};
+
+enum Mode {
+    /// In static mode this pipe will connect to a well-known peer, or
+    /// receive _one_ connection on a well-known address, or does
+    /// both at the same time
+    Static,
+    /// In dynamic mode this pipe connects to a peer via an
+    /// initialisation handshake, or listens for initialisation
+    /// handshakes
+    Dynamic,
+}
+
+/// A builder structure for pipes
+///
+/// The easiest way to create a pipe is via
+/// `builder.basic().static()`.
+pub struct PipeBuilder {
+    send_hooks: SystemBuilder<Context, OckamMessage>,
+    recv_hooks: SystemBuilder<Context, OckamMessage>,
+    /// The selected pipe initialisation mode
+    mode: Mode,
+    /// Peer information
```

Comment:
Why does it have both `peer` and `recv`? Is it bidirectional?

Reply:
The builder is bi-directional, yea. So you use the same builder for both building the sender and receiver